pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer_test.go

```diff
@@ -165,11 +166,335 @@ func TestApplyLocalitySetting(t *testing.T) {
+							{
+								HostIdentifier: buildEndpoint("4.4.4.4"),
+								LoadBalancingWeight: &wrappers.UInt32Value{
+									Value: 1,
+								},
+							},
+						},
+						LoadBalancingWeight: &wrappers.UInt32Value{
+							Value: 2,
+						},
+						Priority: 0,
+					},
+				},
+			},
+			{
+				name:             "match network label",
+				failoverPriority: []string{"topology.istio.io/network", "topology.istio.io/cluster"},
+				proxyLabels: map[string]string{
+					"topology.istio.io/network": "n1",
+					"topology.istio.io/cluster": "test",
+				},
+				expected: []*endpoint.LocalityLbEndpoints{
+					{
+						Locality: &core.Locality{
+							Region:  "region1",
+							Zone:    "zone1",
+							SubZone: "subzone1",
+						},
+						LbEndpoints: []*endpoint.LbEndpoint{
+							{
+								HostIdentifier: buildEndpoint("1.1.1.1"),
+								LoadBalancingWeight: &wrappers.UInt32Value{
+									Value: 1,
+								},
+							},
+						},
+						LoadBalancingWeight: &wrappers.UInt32Value{
+							Value: 1,
+						},
+						Priority: 0,
+					},
+					{
+						Locality: &core.Locality{
+							Region:  "region1",
+							Zone:    "zone1",
+							SubZone: "subzone1",
+						},
+						LbEndpoints: []*endpoint.LbEndpoint{
+							{
+								HostIdentifier: buildEndpoint("2.2.2.2"),
+								LoadBalancingWeight: &wrappers.UInt32Value{
+									Value: 1,
+								},
+							},
+						},
+						LoadBalancingWeight: &wrappers.UInt32Value{
+							Value: 1,
+						},
+						Priority: 1,
+					},
+					{
+						Locality: &core.Locality{
+							Region:  "region2",
+							Zone:    "zone2",
+							SubZone: "subzone2",
+						},
+						LbEndpoints: []*endpoint.LbEndpoint{
+							{
+								HostIdentifier: buildEndpoint("3.3.3.3"),
+								LoadBalancingWeight: &wrappers.UInt32Value{
+									Value: 1,
+								},
+							},
+						},
+						LoadBalancingWeight: &wrappers.UInt32Value{
+							Value: 1,
+						},
+						Priority: 0,
+					},
+					{
+						Locality: &core.Locality{
+							Region:  "region2",
+							Zone:    "zone2",
+							SubZone: "subzone2",
+						},
+						LbEndpoints: []*endpoint.LbEndpoint{
+							{
+								HostIdentifier: buildEndpoint("4.4.4.4"),
+								LoadBalancingWeight: &wrappers.UInt32Value{
+									Value: 1,
+								},
+							},
+						},
+						LoadBalancingWeight: &wrappers.UInt32Value{
+							Value: 1,
+						},
+						Priority: 1,
+					},
+				},
+			},
+			{
+				name:             "match cluster label",
+				failoverPriority: []string{"topology.istio.io/network", "topology.istio.io/cluster"},
+				proxyLabels: map[string]string{
+					"topology.istio.io/network": "test",
+					"topology.istio.io/cluster": "c1",
+				},
+				expected: []*endpoint.LocalityLbEndpoints{
+					{
+						Locality: &core.Locality{
+							Region:  "region1",
+							Zone:    "zone1",
+							SubZone: "subzone1",
+						},
+						LbEndpoints: []*endpoint.LbEndpoint{
+							{
+								HostIdentifier: buildEndpoint("1.1.1.1"),
```

Comment:
The semantics is: if both `"topology.istio.io/network"` and `"topology.istio.io/cluster"` match, it is p0

if only ``"topology.istio.io/network"`` matches, it is p1

others it is p2.

We could not implement an arbitrary match.

Reply:
Sorry I did not understand. Based on this https://github.com/istio/istio/blob/f8571ad3a69a78a4c5f21544f5c3d7ad50902cfe/pilot/pkg/networking/core/v1alpha3/loadbalancer/loadbalancer_test.go#L926 only cluster matches. why did all endpoints got in to P0 priority?

Reply:
Because envoy requires priority increases one by one[0,1,2], but donot allow [2, 3,4].

Reply:
Ok. So only if all labels match - it is P0. If first label match it is P1 and if only second label matches it is p2.  So if only one endpoint matches second label - all endpoints will end up in P0? Would n't the matching one be in P0 and rest in P1/p2?

Also are region/zone/subzone seems internally considered even though they have not specified?

Can you please add a test with more than 3 labels and see how (N-1) priority works?

Reply:
Ok. So only if all labels match - it is P0. If first label match it is P1 and if only second label matches it is p2.  So if only one endpoint matches second label - all endpoints will end up in P0? Would n't the matching one be in P0 and rest in P1/p2?

Also are region/zone/subzone seems internally considered even though they have not specified?

Can you please add a test with more than 3 labels and see how (N-1) priority works?

Reply:
https://github.com/istio/api/blob/94eb8382161503588acf0cab4bb3bd3c30fe1df0/networking/v1alpha3/destination_rule.proto#L1087-L1090 I think you are misreading it

Reply:
`region/zone/subzone` are not considered unless they are set

Reply:
> Can you please add a test with more than 3 labels and see how (N-1) priority works?

done

Reply:
Assuming that n=2 which is the case here 
1. Endpoints matching both network and cluster label should be in P0 - which is working as expected in test
2. Endpoints matching first n-1 label i.e. only first label which is network is in P1 - which is also working correctly
3. By extension of this logic, endpoints matching only the first label with the client proxy has priority P(N-1) i.e. second lowest priority. - This is also working correctly as second lowest priority is P1.
4. Endpoints matching no labels with the client proxy have priority P(N) i.e. lowest priority. - This is also working correctly.

So if an endpoint matches nth label i.e. second label here which is cluster - It is not documented how it will behave. So it is not clear to me why they should not be given high priority than the ones which does not match any labels? Any specific reason? Or Am I missing some thing?

Reply:
@ramaraochavali one way to make it make more sense:

for region/zone: us-west1/zone-a, eu-east1/zone-a.

A match on `zone-a` without region does not make sense at all

Reply:
Yes, maybe the api docs (`Endpoints matching no labels with the client proxy have priority P(N) i.e. lowest priority.`) should be updated to `Endpoints do not match the first 1 label with the client proxy have priority P(N) i.e. lowest priority.`

Reply:
@howardjohn makes sense now

@hzxuzhonghu to make it more clear we may have to say "for a label to be considered for match, the previous label should match i.e. nth label would be considered only if n-1 label matches"?